### PR TITLE
Implement LoRA training and UI for training and UI for inference in i…

### PR DESCRIPTION
**Pull request by KyleHerndon**
…mg2img, inpaint, outpaint

txt2img inference UI is already committed.

### CODE REVIEW

File: apps/stable_diffusion/src/pipelines/pipeline_shark_stable_diffusion_utils.py [hunk 1]
```diff
@@ -360,6 +360,8 @@ def from_pretrained(
 
             clip, unet, vae = mlir_import()
             return cls(vae, clip, get_tokenizer(), unet, scheduler)
+        elif use_lora != "":
+            print("Warning: LoRA specified but you must use '--import-mlir' for this to take effect")
```
Comment (powderluv) [inline apps/stable_diffusion/src/pipelines/pipeline_shark_stable_diffusion_utils.py:364]: should we auto set import-mlir here and just say we are setting it that way ?

File: apps/stable_diffusion/web/index.py [hunk 2]
```diff
@@ -93,6 +94,8 @@ def register_button_click(button, selectedid, inputs, outputs):
             outpaint_web.render()
         with gr.TabItem(label="Upscaler", id=4):
             upscaler_web.render()
+        with gr.TabItem(label="Lora Training", id=5):
```
[the remark this answers is not in the document]
Comment (cstueckrath) [reply apps/stable_diffusion/web/index.py:97]: Training
You want multiple options eventually?

File: apps/stable_diffusion/src/pipelines/pipeline_shark_stable_diffusion_txt2img.py [hunk 3]
```diff
@@ -39,8 +39,9 @@ def __init__(
             SharkEulerDiscreteScheduler,
             DEISMultistepScheduler,
         ],
+        lora_path: str = "",
     ):
-        super().__init__(vae, text_encoder, tokenizer, unet, scheduler)
+        super().__init__(vae, text_encoder, tokenizer, unet, scheduler, lora_path)
```
Comment (Abhishek-Varma) [inline apps/stable_diffusion/src/pipelines/pipeline_shark_stable_diffusion_txt2img.py:44]: I've a couple of questions here for my understanding :-

I just checked the super class `StableDiffusionPipeline` and it has no `lora_path` argument as part of its `__init__` or even a member variable expecting one.

So, why do we require this change in `txt2img` pipeline?

Also, I believe this change would bail out at the interpreter/compiler in the first place. Was this tested?

**Review by Abhishek-Varma — CHANGES_REQUESTED** [review]

File: apps/stable_diffusion/web/ui/outpaint_ui.py [hunk 4]
```diff
@@ -65,6 +65,23 @@
                     label="Input Image", type="pil"
                 ).style(height=300)
 
+                with gr.Accordion(
+                    label="Lora based inference option", open=False
```
[the remark this answers is not in the document]
Comment (KyleHerndon) [reply apps/stable_diffusion/web/ui/outpaint_ui.py]: I made it "LoRA Options." In the UI it looks really strange that it's adjacent and equivalent in style to "Advanced Options" and "Stencil Options" without itself being "Foo Options"

**Comment by KyleHerndon** [comment]
Alright, this is now tested and ready.

**Comment by powderluv** [comment]
We still would need @vivekkhandelwal1 torch-mlir patches to land right ?

**Comment by KyleHerndon** [comment]
There are instructions for how to get it running by using a local branch of torch-mlir. If you want to wait for his patches to land, that's fine too.

**Comment by powderluv** [comment]
No lets land this. That can come. Please validated after this lands the normal webui flows are not broken.

**Pull request merged**